x-pack/plugins/lens/public/indexpattern_datasource/datapanel.tsx

```diff
@@ -257,48 +257,59 @@ export const InnerIndexPatternDataPanel = function InnerIndexPatternDataPanel({
     );
   };
 
-  const allSupportedTypesFields = allFields.filter((field) => supportedFieldTypes.has(field.type));
-
-  const fieldGroups: FieldsGroup = {
-    ...defaultFieldGroups,
-    ...groupBy(allSupportedTypesFields, (field) => {
-      if (field.type === 'document') {
-        return 'specialFields';
-      } else if (containsData(field)) {
-        return 'availableFields';
-      } else return 'emptyFields';
-    }),
-  };
-
-  const filterFieldGroup = (fieldGroup: IndexPatternField[]) =>
-    fieldGroup.filter((field) => {
-      if (
-        localState.nameFilter.length &&
-        !field.name.toLowerCase().includes(localState.nameFilter.toLowerCase())
-      ) {
-        return false;
-      }
+  const fieldGroups: FieldsGroup = useMemo(() => {
+    const allSupportedTypesFields = allFields.filter((field) =>
+      supportedFieldTypes.has(field.type)
+    );
+    const sorted = allSupportedTypesFields.sort(sortFields);
+    // optimization before existingFields are synced
+    if (!hasSyncedExistingFields) {
+      return {
+        ...defaultFieldGroups,
+        ...groupBy(sorted, (field) => {
+          if (field.type === 'document') {
+            return 'specialFields';
+          } else {
+            return 'emptyFields';
+          }
+        }),
+      };
+    }
+    return {
+      ...defaultFieldGroups,
+      ...groupBy(sorted, (field) => {
+        if (field.type === 'document') {
+          return 'specialFields';
+        } else if (containsData(field)) {
+          return 'availableFields';
+        } else return 'emptyFields';
+      }),
+    };
+  }, [allFields, existingFields[currentIndexPattern.title], currentIndexPattern]);
```

Comment:
I think that if we enable the linting rules for hooks, this would complain about missing an explicit dependency on `hasSyncedExistingFields` and `defaultFieldGroups`